rpc/src/rpc_health.rs

```diff
@@ -54,84 +52,74 @@ impl RpcHealth {
-                    // `health_check_slot_distance` of the latest known validator's account hash slot
-                    if latest_account_hash_slot
-                        > latest_known_validator_account_hash_slot
-                            .saturating_sub(self.health_check_slot_distance)
-                    {
-                        RpcHealthStatus::Ok
-                    } else {
-                        let num_slots = latest_known_validator_account_hash_slot
-                            .saturating_sub(latest_account_hash_slot);
-                        warn!(
-                            "health check: behind by {} slots: me={}, latest known_validator={}",
-                            num_slots,
-                            latest_account_hash_slot,
-                            latest_known_validator_account_hash_slot
-                        );
-                        RpcHealthStatus::Behind { num_slots }
-                    }
-                }
-                (latest_account_hash_slot, latest_known_validator_account_hash_slot) => {
-                    if latest_account_hash_slot.is_none() {
-                        warn!("health check: latest_account_hash_slot not available");
-                    }
-                    if latest_known_validator_account_hash_slot.is_none() {
-                        warn!(
-                            "health check: latest_known_validator_account_hash_slot not available"
-                        );
-                    }
-                    RpcHealthStatus::Unknown
-                }
+        // A node can observe votes by both replaying blocks and observing gossip.
+        //
+        // ClusterInfoVoteListener receives votes from both of these sources and then records
+        // optimistically confirmed slots in the Blockstore via OptimisticConfirmationVerifier.
+        // Thus, it is possible for a node to record an optimistically confirmed slot before the
+        // node has replayed and validated the slot for itself.
+        //
+        // OptimisticallyConfirmedBank holds a bank for the latest optimistically confirmed slot
+        // that the node has replayed. It is true that the node will have replayed that slot by
+        // virtue of having a bank available. Observing that the cluster has optimistically
+        // confirmed a slot through gossip is not enough to reconstruct the bank.
+        //
+        // So, comparing the latest optimistic slot from the Blockstore vs. the slot from the
+        // OptimisticallyConfirmedBank bank allows a node to see where it stands in relation to the
+        // tip of the cluster.
+        let my_latest_optimistically_confirmed_slot = self
+            .optimistically_confirmed_bank
+            .read()
+            .unwrap()
+            .bank
+            .slot();
+
+        let mut optimistic_slot_infos = match self.blockstore.get_latest_optimistic_slots(1) {
```

Comment:
I think you could use an `.expect()` here, looks like this only returns `Ok(...)`. Or the API could be changed to remove the `Result`?

Reply:
That function technically returns a result; it could fail if getting the iterator fails
https://github.com/solana-labs/solana/blob/09e858d93971ccc768080fd5b658ecb7c8570908/ledger/src/blockstore.rs#L3145-L3152

This should technically only happen if something is VERY wrong, but in any case, leaned with avoiding the `.expect()` ... don't want the health check to panic and bring the whole process down

Reply:
Getting the iterator won't fail:

https://github.com/solana-labs/solana/blob/09e858d93971ccc768080fd5b658ecb7c8570908/ledger/src/blockstore.rs#L559-L567

and:

https://github.com/solana-labs/solana/blob/09e858d93971ccc768080fd5b658ecb7c8570908/ledger/src/blockstore_db.rs#L1343-L1356

Not sure of the history. I see other functions in `blockstore_db.rs` returning `Result` which won't fail.

Having error handling rather than expect is safer if lower level functions change such that they can fail. It's somewhat ugly to add unreachable error handling code, but this is probably already the case for consumers of these functions.

Reply:
Huh, you're right. I thought we may have had an `.unwrap()` in the iterator codepath somewhere, but it looks like the rust rocks-wrapper doesn't return a `Result`:
https://docs.rs/rocksdb/latest/rocksdb/struct.DBCommon.html#method.iterator_cf

This is inconsistent with other functions that return a `Result` like get or put:
https://docs.rs/rocksdb/latest/rocksdb/struct.DBCommon.html#method.get_cf
https://docs.rs/rocksdb/latest/rocksdb/struct.DBCommon.html#method.put_cf

My guess is we went for consistency across the `Blockstore` API in having (almost) everything return `Result`.

As for this function, I'm a little torn:
- I agree with your statement about having error handling high up being good incase the underlying impl changes such that it could fail (given RockDB's maturity, doubt this actually happens)
- Also agree with the sentiment about unnecessary error handling

I think I'm inclined to leave the error handling in place for now to opt on safe side with my first bullet above. Removing the error handling at the call-site is dependent on intimate knowledge that the underlying API can't raise an error. If we were so inclined, I think the proper change would be to adjust the `Blockstore` API to strip off the `Result`, and update all callers at that point. Does this sound reasonable ?

Reply:
I guess this API is abstracting the underlying database implementation. So it's easy to change RocksDB 😉